lib/resty/upstream/healthcheck.lua

```diff
@@ -214,64 +211,53 @@ local function check_peer(ctx, id, peer, is_backup)
         return
     end
 
+    -- define shortcut error function
+    local report_error = function(...) 
+      if not peer.down then errlog(...) end
+      sock:close()
+      peer_fail(ctx, is_backup, id, peer) 
+    end
+    
     sock:settimeout(ctx.timeout)
 
     if peer.host then
-        -- print("peer port: ", peer.port)
         ok, err = sock:connect(peer.host, peer.port)
     else
         ok, err = sock:connect(name)
     end
     if not ok then
-        if not peer.down then
-            errlog("failed to connect to ", name, ": ", err)
+        return report_error("failed to connect to ", name, ": ", err)
```

Comment:
Oh, ignore this comment since I didn't notice that you do the peer.down check inside the `report_error` function. But as I said, creating closures on such hot code paths is too expensive.